homeassistant/components/recorder/__init__.py

```diff
@@ -195,31 +209,41 @@ def __init__(self, hass: HomeAssistant, purge_days: int, uri: str,
 
         def start_recording(event):
             """Start recording."""
-            self.start()
+            self.start_recording.set()
 
         hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_recording)
         hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
         hass.bus.listen(MATCH_ALL, self.event_listener)
 
+        self.get_session = None
+
+        self.start()
+
     def run(self):
         """Start processing events to save."""
         from homeassistant.components.recorder.models import Events, States
-        import sqlalchemy.exc
+        from homeassistant.components.recorder.restore_state import (
+            load_restore_cache)
+        from sqlalchemy.exc import SQLAlchemyError
 
         while True:
             try:
                 self._setup_connection()
                 self._setup_run()
                 break
-            except sqlalchemy.exc.SQLAlchemyError as err:
+            except SQLAlchemyError as err:
                 _LOGGER.error("Error during connection setup: %s (retrying "
                               "in %s seconds)", err, CONNECT_RETRY_WAIT)
                 time.sleep(CONNECT_RETRY_WAIT)
 
+        load_restore_cache(self.hass)
```

Comment:
We should remove this here and instead just initialize it the first time `get_last_state` is called. Cache should only be initialized if `hass.state == CoreState.starting` (CoreState can be imported from `homeassistant.core`)